drivers/net/netvsc/hn_ethdev.c

```diff
@@ -1129,12 +1098,146 @@ hn_detach(struct hn_data *hv)
 	hn_rndis_detach(hv);
 }
 
+static int
+hn_dev_mtu_set(struct rte_eth_dev *dev, uint16_t mtu)
+{
+	struct hn_data *hv = dev->data->dev_private;
+	struct hn_rx_queue **rxqs = (struct hn_rx_queue **)dev->data->rx_queues;
+	struct hn_tx_queue **txqs = (struct hn_tx_queue **)dev->data->tx_queues;
+	struct rte_eth_dev *vf_dev;
+	unsigned int orig_mtu = dev->data->mtu;
+	uint32_t rndis_mtu;
+	int ret = 0;
+	int i;
+
+	if (dev->data->dev_started) {
+		PMD_DRV_LOG(ERR, "Device must be stopped before changing MTU");
+		return -EIO;
+	}
+
+	/* Change MTU of underlying VF netdev first, if it exists */
+	rte_rwlock_read_lock(&hv->vf_lock);
+	vf_dev = hn_get_vf_dev(hv);
+	if (hv->vf_ctx.vf_vsc_switched && vf_dev) {
+		ret = vf_dev->dev_ops->mtu_set(vf_dev, mtu);
+		if (ret) {
+			rte_rwlock_read_unlock(&hv->vf_lock);
+			goto out;
+		}
+	}
+	rte_rwlock_read_unlock(&hv->vf_lock);
+
+	/* Release channel resources */
+	hn_detach(hv);
+
+	/* Close vmbus channels */
+	for (i = 0; i < hv->num_queues; i++)
+		rte_vmbus_chan_close(hv->channels[i]);
```

Comment:
Need to deal separately with the primary channel, as rte_vmbus_chan_close() doesn't delete it

Reply:
This function is currently used in 3 other places: in 2 of those 3, it is directly passed the primary channel. That makes it a no-op in those cases as well as the 0 index case here, as you point out. I could either (1) just call `rte_free` for the primary channel here, or (2) modify `rte_vmbus_chan_close` to handle closing the primary channel as well. Which do you prefer?

Reply:
Ended up just calling `rte_free` for the primary channel here.